### PR TITLE
[PWA-811] Provide ability to translate Venia's header / footer

### PR DESCRIPTION
**Pull request by anthoula**
## Description

Provide the ability to store translation strings locally, pull correct locale from provided store view code, and translate the header and footer.

## Related Issue
https://jira.corp.magento.com/browse/PWA-811

### Verification Stakeholders
@jimbo 
@dpatil-magento 

### Verification Steps

1. Create new store in Magento with different language (e.g. French)
2. Update env variable to match store view code
3. Add a json file with translation strings in the `i18n` directory
4. Ensure strings in the header and footer are translated

## Checklist

- [x] I have added tests to cover my changes, if necessary.
- [x] I have updated the documentation accordingly, if necessary.

### CODE REVIEW

File: packages/pwa-buildpack/lib/Utilities/graphQL.js [hunk 1]
```diff
@@ -61,7 +62,7 @@ const getUnionAndInterfaceTypes = () => {
 };
 
 module.exports = {
-    getMediaURL,
+    getStoreConfigData,
```
Comment (davemacaulay) [inline packages/pwa-buildpack/lib/Utilities/graphQL.js:65]: Is there a way of avoiding this breaking change? Might be better to retain the `getMediaURL` and include the `getStoreConfigData`.

File: packages/venia-concept/src/index.js [hunk 2]
```diff
@@ -20,6 +21,21 @@ import { registerSW } from './registerSW';
...
+let messages;
+
+try {
+    messages = require(`@magento/venia-ui/lib/i18n/${language.replace(
+        '-',
+        '_'
+    )}.json`);
+} catch (err) {
+    console.warn(
+        `Unable to load translation file. Loading 'en_US' instead. \n${err}`
```
Comment (davemacaulay) [inline packages/venia-concept/src/index.js:34]: I would be inclined to include which file path was attempted to be loaded here to help with debugging.

File: packages/venia-concept/src/index.js [hunk 3]
```diff
@@ -110,10 +127,25 @@ const apolloLink = ApolloLink.from([
     Adapter.apolloLink(apiBase)
 ]);
 
+const onIntlError = error => {
+    if (error.code === 'MISSING_TRANSLATION') {
+        console.warn('Missing translation', error.message);
```
Comment (davemacaulay) [inline packages/venia-concept/src/index.js:132]: Does `error.message` include sufficient information here? or do we need to supplement it? :)

File: packages/venia-concept/src/index.js [hunk 4]
```diff
@@ -20,6 +21,21 @@ import { registerSW } from './registerSW';
...
 const apiBase = new URL('/graphql', location.origin).toString();
 
+const language = STORE_VIEW_LOCALE;
+let messages;
+
+try {
+    messages = require(`@magento/venia-ui/lib/i18n/${language.replace(
+        '-',
+        '_'
+    )}.json`);
```
Comment (davemacaulay) [inline packages/venia-concept/src/index.js:31]: What do you think about making 2 utility functions? one to covert from Magento to `react-intl` and one to convert it back?

File: packages/venia-ui/lib/components/Header/__tests__/__snapshots__/cartTrigger.spec.js.snap [hunk 5]
```diff
@@ -4,7 +4,7 @@ exports[`Cart counter displays 99+ when items quantity is more than 99 1`] = `
 Array [
   <div>
     <button
-      aria-label="Toggle mini cart. You have 100 items in your cart."
+      aria-label="Toggle mini cart. You have {count} items in your cart."
```
Comment (davemacaulay) [inline packages/venia-ui/lib/components/Header/__tests__/__snapshots__/cartTrigger.spec.js.snap:7]: Would we not expect for the test runner to convert the {count} variable into the actual integer here?

File: packages/venia-ui/lib/i18n/uk_UA.json [hunk 6]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (davemacaulay) [inline packages/venia-ui/lib/i18n/uk_UA.json:1]: I love that you translated this to French and Ukrainian!

File: packages/pwa-buildpack/lib/Utilities/graphQL.js (hunk 1 above)
Comment (larsroettig) [reply packages/pwa-buildpack/lib/Utilities/graphQL.js:65]: I think we should think about https://github.com/magento/pwa-studio/issues/2120 if we need a breaking change also all developer will benefit oft this

**Review by larsroettig — CHANGES_REQUESTED** [review]

File: packages/venia-concept/src/index.js (hunk 3 above)
Comment (anthoula) [reply packages/venia-concept/src/index.js:135]: Error messaging is pretty clear:
`Missing translation [React Intl Error MISSING_TRANSLATION] Missing message: "About Us" for locale "fr-FR", using id as fallback`

File: packages/pwa-buildpack/lib/Utilities/graphQL.js (hunk 1 above)
Comment (anthoula) [reply packages/pwa-buildpack/lib/Utilities/graphQL.js]: @davemacaulay Yes, getMediaURL has been restored

Comment (anthoula) [reply packages/pwa-buildpack/lib/Utilities/graphQL.js]: @larsroettig Thanks for pointing that out

File: jest.config.js [hunk 7]
```diff
@@ -189,7 +189,8 @@ const testReactComponents = inPackage => ({
                 ]
             }
         },
-        STORE_NAME: 'Venia'
+        STORE_NAME: 'Venia',
+        STORE_VIEW_LOCALE: 'en-US'
```
Comment (supernova-at) [inline jest.config.js:193]: Should this be `STORE_VIEW_CODE`?

File: packages/venia-concept/src/index.js [hunk 8]
```diff
@@ -16,10 +17,28 @@ import app from '@magento/peregrine/lib/store/actions/app';
...
+try {
+    messages = require(`@magento/venia-ui/lib/i18n/${fromReactIntl(
+        language
+    )}.json`);
+} catch (err) {
+    console.warn(
+        `Unable to load translation file. Loading 'venia-ui/lib/i18n/en_US.json' instead. \n${err}`
+    );
+    messages = require(`@magento/venia-ui/lib/i18n/en_US.json`);
+}
```
Comment (jimbo) [inline packages/venia-concept/src/index.js:40]: I believe we should use [dynamic import](https://webpack.js.org/api/module-methods/#dynamic-expressions-in-import) statements in the browser, rather than `require` statements. Note that these are asynchronous and that we can do some cool things with them, such as preloading.

Since I'm recommending that we switch from a synchronous (blocking) require to an async (non-blocking) one, that means we'll have to re-render `IntlProvider` when the call completes and we have real `messages` to give it. As a result, we should probably move this dynamic import into a component that wraps `IntlProvider`.

Finally, we should do as much of this as possible in `venia-ui` rather than `venia-concept`, I think. Making a component would be a good way to do that.

```es6
// shouldn't even have to provide a default, right?
// https://formatjs.io/docs/react-intl/components#defaultlocale-and-defaultformats
const locale = fromReactIntl(language)

const LocaleProvider = props => {
  const { children } = props
  const [messages, setMessages] = useState(null)

  // try doing the dynamic import as an effect
  // and set the result in state, so we can render before & after
  useEffect(
    () => {
      import(`@magento/venia-ui/lib/i18n/${locale}.json`)
        .then(data => { setMessages(data) })
        .catch(error => { console.error("Unable to load translation file.") })
    },
    [setMessages]
  )

  return (
    <IntlProvider
      key={language}
      {...props}
      defaultLocale="en-US"
      locale={language}
      messages={messages}
    />
  )
}

File: packages/venia-ui/lib/components/AccountMenu/accountMenuItems.js [hunk 9]
```diff
@@ -25,6 +16,25 @@ const AccountMenuItems = props => {
...
+        },
+        { name: intl.formatMessage({ id: 'Favorites Lists' }), url: '' },
+        { name: intl.formatMessage({ id: 'Address Book' }), url: '' },
+        { name: intl.formatMessage({ id: 'Saved Payments' }), url: '' },
+        {
+            name: intl.formatMessage({ id: 'Communications' }),
+            url: '/communications'
+        },
+        { name: intl.formatMessage({ id: 'Account Information' }), url: '' }
+    ];
```
Comment (jimbo) [inline packages/venia-ui/lib/components/AccountMenu/accountMenuItems.js:37]: Yes, we do need to translate these, but we should probably continue to define `MENU_ITEMS` statically (before the component) and just call `formatMessage` from the same map operation we're already doing to create elements.

We might also take this opportunity to give these static items keys that more uniquely identify them within the app (e.g., `accountMenu.orderHistoryLink`), since that kinda seems to be what `react-intl` expects us to do. Thoughts?

```es6
const MENU_ITEMS = [/* ... */]

const AccountMenuItems = props => {
  // ...

  const { formatMessage } = useIntl()

  // could call `intl.formatMessage(item.name)` on each
  // but we might as well just use `FormattedMessage`, no?
  const menuItems = MENU_ITEMS.map(item => (
    <Link key={item.name} className={classes.link} to={item.url}>
      <FormattedMessage id={item.name} defaultMessage={item.name} />
    </Link>
  )
}
```

File: packages/venia-ui/lib/components/AccountMenu/accountMenuItems.js [hunk 10]
```diff
@@ -26,9 +27,12 @@ const AccountMenuItems = props => {
     const classes = mergeClasses(defaultClasses, props.classes);
 
     const menuItems = MENU_ITEMS.map(item => {
+        const id = `account_menu.${item.name
+            .toLowerCase()
+            .replace(/ /g, '_')}_link`;
```
Comment (jimbo) [inline packages/venia-ui/lib/components/AccountMenu/accountMenuItems.js:32]: This works, but I think I'd rather just add an `id` property to each object in `MENU_ITEMS` rather than generating it here. Also, I think we should probably move `MENU_ITEMS` to the talon, and return it (under the property `menuItems`). Would that work?

Looks much better with `<FormattedMessage id={id} />` though. 👍

File: packages/venia-ui/lib/components/AccountMenu/__tests__/__snapshots__/accountMenuItems.spec.js.snap [hunk 11]
```diff
@@ -2,13 +2,13 @@
 
 exports[`it renders correctly 1`] = `
 <div>
-  &lt;Link&gt;Order History&lt;/Link&gt;
```
Comment (davemacaulay) [inline packages/venia-ui/lib/components/AccountMenu/__tests__/__snapshots__/accountMenuItems.spec.js.snap:5]: Is this change intentional? Should we not still have a string represented in the snapshot?

File: packages/venia-ui/lib/components/AccountMenu/__tests__/__snapshots__/accountMenuItems.spec.js.snap [hunk 12]
```diff
@@ -2,13 +2,13 @@
...
-  &lt;Link&gt;Saved Payments&lt;/Link&gt;
-  &lt;Link&gt;Communications&lt;/Link&gt;
-  &lt;Link&gt;Account Information&lt;/Link&gt;
+  &lt;Link&gt;[object Object]&lt;/Link&gt;
+  &lt;Link&gt;[object Object]&lt;/Link&gt;
+  &lt;Link&gt;[object Object]&lt;/Link&gt;
+  &lt;Link&gt;[object Object]&lt;/Link&gt;
+  &lt;Link&gt;[object Object]&lt;/Link&gt;
+  &lt;Link&gt;[object Object]&lt;/Link&gt;
+  &lt;Link&gt;[object Object]&lt;/Link&gt;
```
Comment (davemacaulay) [inline packages/venia-ui/lib/components/AccountMenu/__tests__/__snapshots__/accountMenuItems.spec.js.snap:11]: Is this change intentional? Should we not still have a string represented in the snapshot?

File: packages/venia-ui/lib/components/Header/__tests__/__snapshots__/cartTrigger.spec.js.snap (hunk 5 above)
Comment (davemacaulay) [reply packages/venia-ui/lib/components/Header/__tests__/__snapshots__/cartTrigger.spec.js.snap:7]: I guess this comment got lost, would be good to understand if the snapshots could contain the replacement patterns or if they should be rendered correctly. As we're mainly writing unit tests I would imagine we'd wanna mock our `react-intl` whenever we do not intend for it to translate.

File: packages/venia-ui/lib/components/Header/accountTrigger.js [hunk 13]
```diff
@@ -28,17 +29,20 @@ const AccountTrigger = props => {
 
     const classes = mergeClasses(defaultClasses, props.classes);
     const rootClassName = accountMenuIsOpen ? classes.root_open : classes.root;
+    const intl = useIntl();
```
Comment (davemacaulay) [inline packages/venia-ui/lib/components/Header/accountTrigger.js:32]: `const { formatMessage } = useIntl();` would be a lot nicer here

Comment (davemacaulay) [reply packages/venia-ui/lib/components/Header/accountTrigger.js:32]: and probably in all other instances unless we're consuming other aspects of `useIntl` frequently.

File: packages/venia-ui/lib/components/AccountMenu/__tests__/__snapshots__/accountMenuItems.spec.js.snap (hunk 12 above)
Comment (jimbo) [reply packages/venia-ui/lib/components/AccountMenu/__tests__/__snapshots__/accountMenuItems.spec.js.snap:11]: I noticed this. I do think we need to mock something in order for snapshots to be accurate. The text replacement logic in `react-intl` components seems pretty robust, so this isn't too surprising.

**Review by jimbo — DISMISSED** [review]

**Comment by dpatil-magento** [comment]
@anthoula please check below observations - 

1. Fresh clone repo and checkout pr branch > yarn install > yarn buildpack create-env-file packages/venia-concept >>  .env >> check - STORE_VIEW_CODE=french  instead of default.
2. Sign in to account > Top right Continuous loading spinner, left drawer still shows login screen and account links are not translated.
image

3. Unable to add product to Cart > Graphql Error -  ‘Wrong store code specified for cart <id>’
4. In Admin UI even when French Store view is not default, how its able to fetch French store view?
image

**Comment by dpatil-magento** [comment]
@anthoula @davemacaulay 
> @anthoula please check below observations -
> 
> 1. Fresh clone repo and checkout pr branch > yarn install > yarn buildpack create-env-file packages/venia-concept >>  .env >> check - STORE_VIEW_CODE=french  instead of default.

Looks good. I think now we should be good to remove entry from prod.dockerfile and dev.dockerfile ? Also, nice to update comment `# Default when not set: ` to `# Default when not set: default` ? 

Rest all previous observations are working.

8. There is change in yarn.lock when I install this branch, just need to commit it? pls check.
